man/csvsql.1
.\" Man page generated from reStructuredText.
.
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.TH "CSVSQL" "1" "Aug 16, 2024" "2.2.0" "csvkit"
.SH NAME
csvsql \- csvsql Documentation
.SH DESCRIPTION
.sp
Generate SQL statements for a CSV file or execute those statements directly on a database. In the latter case supports both creating tables and inserting data:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
usage: csvsql [\-h] [\-d DELIMITER] [\-t] [\-q QUOTECHAR] [\-u {0,1,2,3}] [\-b]
              [\-p ESCAPECHAR] [\-z FIELD_SIZE_LIMIT] [\-e ENCODING] [\-L LOCALE]
              [\-S] [\-\-blanks] [\-\-null\-value NULL_VALUES [NULL_VALUES ...]]
              [\-\-date\-format DATE_FORMAT] [\-\-datetime\-format DATETIME_FORMAT]
              [\-H] [\-K SKIP_LINES] [\-v] [\-l] [\-\-zero] [\-V]
              [\-i {firebird,mssql,mysql,oracle,postgresql,sqlite,sybase}]
              [\-\-db CONNECTION_STRING] [\-\-query QUERIES] [\-\-insert]
              [\-\-prefix PREFIX] [\-\-before\-insert BEFORE_INSERT]
              [\-\-after\-insert AFTER_INSERT] [\-\-tables TABLE_NAMES]
              [\-\-no\-constraints] [\-\-unique\-constraint UNIQUE_CONSTRAINT]
              [\-\-no\-create] [\-\-create\-if\-not\-exists] [\-\-overwrite]
              [\-\-db\-schema DB_SCHEMA] [\-y SNIFF_LIMIT] [\-I]
              [\-\-chunk\-size CHUNK_SIZE]
              [FILE [FILE ...]]

Generate SQL statements for one or more CSV files, or execute those statements
directly on a database, and execute one or more SQL queries.

positional arguments:
  FILE                  The CSV file(s) to operate on. If omitted, will accept
                        input as piped data via STDIN.

optional arguments:
  \-h, \-\-help            show this help message and exit
  \-i {mssql,mysql,oracle,postgresql,sqlite,duckdb,crate,ingres}, \-\-dialect {mssql,mysql,oracle,postgresql,sqlite,duckdb,crate,ingres}
                        Dialect of SQL to generate. Cannot be used with \-\-db.
  \-\-db CONNECTION_STRING
                        If present, a SQLAlchemy connection string to use to
                        directly execute generated SQL on a database.
  \-\-engine\-option ENGINE_OPTION ENGINE_OPTION
                        A keyword argument to SQLAlchemy\(aqs create_engine(), as
                        a space\-separated pair. This option can be specified
                        multiple times. For example: thick_mode True
  \-\-query QUERIES       Execute one or more SQL queries delimited by \-\-sql\-
                        delimiter, and output the result of the last query as
                        CSV. QUERY may be a filename. \-\-query may be specified
                        multiple times.
  \-\-insert              Insert the data into the table. Requires \-\-db.
  \-\-prefix PREFIX       Add an expression following the INSERT keyword, like
                        OR IGNORE or OR REPLACE.
  \-\-before\-insert BEFORE_INSERT
                        Before the INSERT command, execute one or more SQL
                        queries delimited by \-\-sql\-delimiter. Requires
                        \-\-insert.
  \-\-after\-insert AFTER_INSERT
                        After the INSERT command, execute one or more SQL
                        queries delimited by \-\-sql\-delimiter. Requires
                        \-\-insert.
  \-\-sql\-delimiter SQL_DELIMITER
                        Delimiter separating SQL queries in \-\-query, \-\-before\-
                        insert, and \-\-after\-insert.
  \-\-tables TABLE_NAMES  A comma\-separated list of names of tables to be
                        created. By default, the tables will be named after
                        the filenames without extensions or \(dqstdin\(dq.
  \-\-no\-constraints      Generate a schema without length limits or null
                        checks. Useful when sampling big tables.
  \-\-unique\-constraint UNIQUE_CONSTRAINT
                        A column\-separated list of names of columns to include
                        in a UNIQUE constraint.
  \-\-no\-create           Skip creating the table. Requires \-\-insert.
  \-\-create\-if\-not\-exists
                        Create the table if it does not exist, otherwise keep
                        going. Requires \-\-insert.
  \-\-overwrite           Drop the table if it already exists. Requires
                        \-\-insert. Cannot be used with \-\-no\-create.
  \-\-db\-schema DB_SCHEMA
                        Optional name of database schema to create table(s)
                        in.
  \-y SNIFF_LIMIT, \-\-snifflimit SNIFF_LIMIT
                        Limit CSV dialect sniffing to the specified number of
                        bytes. Specify \(dq0\(dq to disable sniffing entirely, or
                        \(dq\-1\(dq to sniff the entire file.
  \-I, \-\-no\-inference    Disable type inference (and \-\-locale, \-\-date\-format,
                        \-\-datetime\-format, \-\-no\-leading\-zeroes) when parsing
                        the input.
  \-\-chunk\-size CHUNK_SIZE
                        Chunk size for batch insert into the table. Requires
                        \-\-insert.
  \-\-min\-col\-len MIN_COL_LEN
                        The minimum length of text columns.
  \-\-col\-len\-multiplier COL_LEN_MULTIPLIER
                        Multiply the maximum column length by this multiplier
                        to accomodate larger values in later runs.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
See also: \fI\%Arguments common to all tools\fP\&.
.sp
For information on connection strings and supported dialects refer to the \fI\%SQLAlchemy documentation\fP\&.
.sp
If you prefer not to enter your password in the connection string, store the password securely in a \fI\%PostgreSQL Password File\fP, a \fI\%MySQL Options File\fP or similar files for other systems.
.sp
\fBNOTE:\fP
.INDENT 0.0
.INDENT 3.5
Using the \fB\-\-query\fP option may cause rounding (in Python 2) or introduce \fI\%Python floating point issues\fP (in Python 3).
.UNINDENT
.UNINDENT
.sp
\fBNOTE:\fP
.INDENT 0.0
.INDENT 3.5
If the CSV file was created from a JSON file using \fI\%in2csv\fP, remember to quote SQL columns properly. For example:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
echo \(aq{\(dqa\(dq:{\(dqb\(dq:\(dqc\(dq},\(dqd\(dq:\(dqe\(dq}\(aq | in2csv \-f ndjson | csvsql \-\-query \(aqSELECT \(dqa/b\(dq FROM stdin\(aq
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.sp
\fBNOTE:\fP
.INDENT 0.0
.INDENT 3.5
Alternatives to \fI\%csvsql\fP are \fI\%q\fP and \fI\%textql\fP\&.
.UNINDENT
.UNINDENT
.SH EXAMPLES
.SS Generate SQL statements
.sp
Generate a statement in the PostgreSQL dialect:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
csvsql \-i postgresql examples/realdata/FY09_EDU_Recipients_by_State.csv
.ft P
.fi
.UNINDENT
.UNINDENT
.SS Interact with a SQL database
.sp
Create a table and import data from the CSV directly into PostgreSQL:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
createdb test
csvsql \-\-db postgresql:///test \-\-tables fy09 \-\-insert examples/realdata/FY09_EDU_Recipients_by_State.csv
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
For large tables it may not be practical to process the entire table. One solution to this is to analyze a sample of the table. In this case it can be useful to turn off length limits and null checks with the \fB\-\-no\-constraints\fP option:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
head \-n 20 examples/realdata/FY09_EDU_Recipients_by_State.csv | csvsql \-\-no\-constraints \-\-tables fy09
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Create tables for an entire directory of CSVs and import data from those files directly into PostgreSQL:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
createdb test
csvsql \-\-db postgresql:///test \-\-insert examples/*_converted.csv
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
If those CSVs have identical headers, you can import them into the same table by using \fI\%csvstack\fP first:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
createdb test
csvstack examples/dummy?.csv | csvsql \-\-db postgresql:///test \-\-insert
.ft P
.fi
.UNINDENT
.UNINDENT
.SS Query and output CSV files using SQL
.sp
You can use csvsql to \(dqdirectly\(dq query one or more CSV files. Please note that this will create an in\-memory SQLite database, so it won\(aqt be very fast:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
csvsql \-\-query  \(dqSELECT m.usda_id, avg(i.sepal_length) AS mean_sepal_length FROM iris AS i JOIN irismeta AS m ON (i.species = m.species) GROUP BY m.species\(dq examples/iris.csv examples/irismeta.csv
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Group rows by one column:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
csvsql \-\-query \(dqSELECT * FROM \(aqdummy3\(aq GROUP BY a\(dq examples/dummy3.csv
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Concatenate two columns:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
csvsql \-\-query \(dqSELECT a || b FROM \(aqdummy3\(aq\(dq \-\-no\-inference examples/dummy3.csv
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
If a column contains null values, you must \fBCOALESCE\fP the column:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
csvsql \-\-query \(dqSELECT a || COALESCE(b, \(aq\(aq) FROM \(aqsort_ints_nulls\(aq\(dq \-\-no\-inference examples/sort_ints_nulls.csv
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The \fBUPDATE\fP SQL statement produces no output. Remember to \fBSELECT\fP the columns and rows you want:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
csvsql \-\-query \(dqUPDATE \(aqdummy3\(aq SET a = \(aqfoo\(aq; SELECT * FROM \(aqdummy3\(aq\(dq examples/dummy3.csv
.ft P
.fi
.UNINDENT
.UNINDENT
.SH AUTHOR
Christopher Groskopf and contributors
.SH COPYRIGHT
2016, Christopher Groskopf and James McKinney
.\" Generated by docutils manpage writer.
.


CHANGELOG
0.9.2
-----

* Fix version of openpyxl.

0.9.1
-----

* Add Antonio Lima to AUTHORS.
* Add support for ndjson. (#329)
* Add missing docs for csvcut -C. (#227)
* Reorganize docs so TOC works better. (#339)
* Render docs locally with RTD theme.
* Fix header in "tricks" docs.
* Add install instructions to tutorial. (#331)
* Add killer examples to doc index. (#328)
* Reorganize doc index
* Fix broken csvkit module documentation. (#327)
* Fix version of openpyxl to work around encoding issue. (#391, #288)

0.9.0
-----

* Write missing sections of the tutorial. (#32)
* Remove -q arg from sql2csv (conflicts with common flag).
* Fix csvjoin in case where left dataset rows without all columns.
* Rewrote tutorial based on LESO data. (#324)
* Don't error in csvjson if lat/lon columns are null. (#326)
* Maintain field order in output of csvjson.
* Add unit test for json in2csv. (#77)
* Maintain key order when converting JSON into CSV. (#325.)
* Upgrade python-dateutil to version 2.2 (#304)
* Fix sorting of columns with null values. (#302)
* Added release documentation.
* Fill out short rows with null values. (#313)
* Fix unicode output for csvlook and csvstat. (#315)
* Add documentation for --zero. (#323)
* Fix Integrity error when inserting zero rows in database with csvsql. (#299)
* Add Michael Mior to AUTHORS. (#305)
* Add --count option to CSVStat.
* Implement csvformat.
* Fix bug causing CSVKitDictWriter to output 'utf-8' for blank fields.

0.8.0
-----

* Add pnaimoli to AUTHORS.
* Fix column specification in csvstat. (#236)
* Added "Tips and Tricks" documentation. (#297, #298)
* Add Espartaco Palma to AUTHORS.
* Remove unnecessary enumerate calls. (#292)
* Deprecated DBF support for Python 3+.
* Add support for Python 3.3 and 3.4 (#239)

0.7.3
-----

* Fix date handling with openpyxl > 2.0 (#285)
* Add Kristina Durivage to AUTHORS. (#243)
* Added Richard Low to AUTHORS.
* Support SQL queries "directly" on CSV files. (#276)
* Add Tasneem Raja to AUTHORS.
* Fix off-by-one error in open ended column ranges. (#238)
* Add Matt Pettis to AUTHORS.
* Add line numbers flag to csvlook (#244)
* Only install argparse for Python < 2.7. (#224)
* Add Diego Rabatone Oliveira to AUTHORS.
* Add Ryan Murphy to AUTHORS.
* Fix DBF dependency. (#270)

0.7.2
-----

* Fix CHANGELOG for release.

0.7.1
-----

* Fix homepage url in setup.py.

0.7.0
-----

* Fix XLSX datetime normalization bug. (#223)
* Add raistlin7447 to AUTHORS.
* Merged sql2csv utility (#259).
* Add Jeroen Janssens to AUTHORS.
* Validate csvsql DB connections before parsing CSVs. (#257)
* Clarify install process for Ubuntu. (#249)
* Clarify docs for --escapechar. (#242)
* Make ``import csvkit`` API compatible with ``import csv``.
* Update Travis CI link. (#258)
* Add Sébastien Fievet to AUTHORS.
* Use case-sensitive name for SQLAlchemy (#237)
* Add Travis Swicegood to AUTHORS.

0.6.1
-----

* Add Chris Rosenthal to AUTHORS.
* Fix multi-file input to csvsql. (#193)
* Passing --snifflimit=0 to disable dialect sniffing. (#190)
* Add aarcro to the AUTHORS file.
* Improve performance of csvgrep. (#204)
* Add Matt Dudys to AUTHORS.
* Add support for --skipinitialspace. (#201)
* Add Joakim Lundborg to AUTHORS.
* Add --no-inference option to in2csv and csvsql. (#206)
* Add Federico Scrinzi to AUTHORS file.
* Add --no-header-row to all tools. (#189)
* Fix csvstack blowing up on empty files. (#209)
* Add Chris Rosenthal to AUTHORS file.
* Add --db-schema option to csvsql. (#216)
* Add Shane StClair to AUTHORS file.
* Add --no-inference support to csvsort. (#222)

0.5.0
-----

* Implement geojson support in csvjson. (#159)
* Optimize writing of eight bit codecs. (#175)
* Created csvpy. (#44)
* Support --not-columns for excluding columns. (#137)
* Add Jan Schulz to AUTHORS file.
* Add Windows scripts. (#111, #176)
* csvjoin, csvsql and csvstack will no longer hold open all files. (#178)
* Added Noah Hoffman to AUTHORS.
* Make csvlook output compatible with emacs table markup. (#174)

0.4.4
-----

* Add Derek Wilson to AUTHORS.
* Add Kevin Schaul to AUTHORS.
* Add DBF support to in2csv. (#11, #160)
* Support --zero option for zero-based column indexing. (#144)
* Support mixing nulls and blanks in string columns.
* Add --blanks option to csvsql. (#149)
* Add multi-file (glob) support to csvsql. (#146)
* Add Gregory Temchenko to AUTHORS.
* Add --no-create option to csvsql. (#148)
* Add Anton Ian Sipos to AUTHORS.
* Fix broken pipe errors. (#150)

0.4.3
-----

* Begin CHANGELOG (a bit late, I'll admit).


csvkit/utilities/csvsql.py
#!/usr/bin/env python
import os.path
import sys

import agate
import agatesql  # noqa: F401
from sqlalchemy import create_engine, dialects

from csvkit.cli import CSVKitUtility, isatty, parse_list

try:
    import importlib_metadata
except ImportError:
    import importlib.metadata as importlib_metadata

DIALECTS = dialects.__all__ + tuple(e.name for e in importlib_metadata.entry_points(group='sqlalchemy.dialects'))


class CSVSQL(CSVKitUtility):
    description = 'Generate SQL statements for one or more CSV files, or execute those statements directly on a ' \
                  'database, and execute one or more SQL queries.'
    # Override 'f' because the utility accepts multiple files.
    override_flags = ['f']

    def add_arguments(self):
        self.argparser.add_argument(
            metavar='FILE', nargs='*', dest='input_paths', default=['-'],
            help='The CSV file(s) to operate on. If omitted, will accept input as piped data via STDIN.')
        self.argparser.add_argument(
            '-i', '--dialect', dest='dialect', choices=DIALECTS,
            help='Dialect of SQL to generate. Cannot be used with --db.')
        self.argparser.add_argument(
            '--db', dest='connection_string',
            help='If present, a SQLAlchemy connection string to use to directly execute generated SQL on a database.')
        self.argparser.add_argument(
            '--engine-option', dest='engine_option', nargs=2, action='append', default=[],
            help="A keyword argument to SQLAlchemy's create_engine(), as a space-separated pair. "
                 "This option can be specified multiple times. For example: thick_mode True")
        self.argparser.add_argument(
            '--query', dest='queries', action='append',
            help='Execute one or more SQL queries delimited by --sql-delimiter, and output the result of the last '
                 'query as CSV. QUERY may be a filename. --query may be specified multiple times.')
        self.argparser.add_argument(
            '--insert', dest='insert', action='store_true',
            help='Insert the data into the table. Requires --db.')
        self.argparser.add_argument(
            '--prefix', action='append', default=[],
            help='Add an expression following the INSERT keyword, like OR IGNORE or OR REPLACE.')
        self.argparser.add_argument(
            '--before-insert', dest='before_insert',
            help='Before the INSERT command, execute one or more SQL queries delimited by --sql-delimiter. '
                 'Requires --insert.')
        self.argparser.add_argument(
            '--after-insert', dest='after_insert',
            help='After the INSERT command, execute one or more SQL queries delimited by --sql-delimiter. '
                 'Requires --insert.')
        self.argparser.add_argument(
            '--sql-delimiter', dest='sql_delimiter', default=';',
            help='Delimiter separating SQL queries in --query, --before-insert, and --after-insert.')
        self.argparser.add_argument(
            '--tables', dest='table_names',
            help='A comma-separated list of names of tables to be created. By default, the tables will be named after '
                 'the filenames without extensions or "stdin".')
        self.argparser.add_argument(
            '--no-constraints', dest='no_constraints', action='store_true',
            help='Generate a schema without length limits or null checks. Useful when sampling big tables.')
        self.argparser.add_argument(
            '--unique-constraint', dest='unique_constraint',
            help='A column-separated list of names of columns to include in a UNIQUE constraint.')
        self.argparser.add_argument(
            '--no-create', dest='no_create', action='store_true',
            help='Skip creating the table. Requires --insert.')
        self.argparser.add_argument(
            '--create-if-not-exists', dest='create_if_not_exists', action='store_true',
            help='Create the table if it does not exist, otherwise keep going. Requires --insert.')
        self.argparser.add_argument(
            '--overwrite', dest='overwrite', action='store_true',
            help='Drop the table if it already exists. Requires --insert. Cannot be used with --no-create.')
        self.argparser.add_argument(
            '--db-schema', dest='db_schema',
            help='Optional name of database schema to create table(s) in.')
        self.argparser.add_argument(
            '-y', '--snifflimit', dest='sniff_limit', type=int, default=1024,
            help='Limit CSV dialect sniffing to the specified number of bytes. '
                 'Specify "0" to disable sniffing entirely, or "-1" to sniff the entire file.')
        self.argparser.add_argument(
            '-I', '--no-inference', dest='no_inference', action='store_true',
            help='Disable type inference (and --locale, --date-format, --datetime-format, --no-leading-zeroes) '
                 'when parsing the input.')
        self.argparser.add_argument(
            '--chunk-size', dest='chunk_size', type=int,
            help='Chunk size for batch insert into the table. Requires --insert.')
        self.argparser.add_argument(
            '--min-col-len', dest='min_col_len', type=int, default=1,
            help='The minimum length of text columns.')
        self.argparser.add_argument(
            '--col-len-multiplier', dest='col_len_multiplier', type=int, default=1,
            help='Multiply the maximum column length by this multiplier to accommodate larger values in later runs.')

    def main(self):
        if isatty(sys.stdin) and self.args.input_paths == ['-']:
            self.argparser.error('You must provide an input file or piped data.')

        self.input_files = []
        self.connection = None
        self.table_names = []
        self.unique_constraint = []

        if self.args.table_names:
            self.table_names = self.args.table_names.split(',')
        if self.args.unique_constraint:
            self.unique_constraint = self.args.unique_constraint.split(',')

        # Create a SQLite database in memory if no connection string is specified
        if self.args.queries and not self.args.connection_string:
            self.args.connection_string = "sqlite:///:memory:"
            self.args.insert = True

        if self.args.dialect and self.args.connection_string:
            self.argparser.error('The --dialect option is only valid when neither --db nor --query are specified.')
        if self.args.insert and not self.args.connection_string:
            self.argparser.error('The --insert option is only valid when either --db or --query is specified.')

        if self.args.no_create and not self.args.insert:
            self.argparser.error('The --no-create option is only valid if --insert is also specified.')
        if self.args.create_if_not_exists and not self.args.insert:
            self.argparser.error('The --create-if-not-exists option is only valid if --insert is also specified.')
        if self.args.overwrite and not self.args.insert:
            self.argparser.error('The --overwrite option is only valid if --insert is also specified.')
        if self.args.overwrite and self.args.no_create:
            self.argparser.error('The --overwrite option is only valid if --no-create is not specified.')
        if self.args.before_insert and not self.args.insert:
            self.argparser.error('The --before-insert option is only valid if --insert is also specified.')
        if self.args.after_insert and not self.args.insert:
            self.argparser.error('The --after-insert option is only valid if --insert is also specified.')
        if self.args.chunk_size and not self.args.insert:
            self.argparser.error('The --chunk-size option is only valid if --insert is also specified.')

        if self.args.no_create and self.args.create_if_not_exists:
            self.argparser.error('The --no-create and --create-if-not-exists options are mutually exclusive.')

        # Lazy open files
        for path in self.args.input_paths:
            self.input_files.append(self._open_input_file(path))

        # Establish database validity before reading CSV files
        if self.args.connection_string:
            try:
                engine = create_engine(self.args.connection_string, **parse_list(self.args.engine_option))
            except ImportError as e:
                raise ImportError(
                    "You don't appear to have the necessary database backend installed for connection string you're "
                    "trying to use. Available backends include:\n\nPostgreSQL:\tpip install psycopg2\nMySQL:\t\tpip "
                    "install mysql-connector-python OR pip install mysqlclient\n\nFor details on connection strings "
                    "and other backends, please see the SQLAlchemy documentation on dialects at:\n\n"
                    "https://www.sqlalchemy.org/docs/dialects/"
                ) from e

            self.connection = engine.connect()

        try:
            self._failsafe_main()
        finally:
            for f in self.input_files:
                f.close()

            if self.connection:
                self.connection.close()
                engine.dispose()

    def _failsafe_main(self):
        """
        Inner main function. If anything fails in here, file handles and
        database connections will be safely closed.
        """
        if self.connection:
            transaction = self.connection.begin()

        for f in self.input_files:
            try:
                # Try to use name specified via --tables
                table_name = self.table_names.pop(0)
            except IndexError:
                if f == sys.stdin:
                    table_name = "stdin"
                else:
                    # Use filename as table name
                    table_name = os.path.splitext(os.path.basename(f.name))[0]

            table = None
            sniff_limit = self.args.sniff_limit if self.args.sniff_limit != -1 else None

            try:
                table = agate.Table.from_csv(
                    f,
                    skip_lines=self.args.skip_lines,
                    sniff_limit=sniff_limit,
                    column_types=self.get_column_types(),
                    **self.reader_kwargs,
                )
            except StopIteration:
                # Catch cases where no table data was provided and fall through
                # to query logic
                continue

            if table:
                if self.connection:
                    if self.args.before_insert:
                        for query in self.args.before_insert.split(self.args.sql_delimiter):
                            self.connection.exec_driver_sql(query)

                    table.to_sql(
                        self.connection,
                        table_name,
                        overwrite=self.args.overwrite,
                        create=not self.args.no_create,
                        create_if_not_exists=self.args.create_if_not_exists,
                        insert=self.args.insert and len(table.rows) > 0,
                        prefixes=self.args.prefix,
                        db_schema=self.args.db_schema,
                        constraints=not self.args.no_constraints,
                        unique_constraint=self.unique_constraint,
                        chunk_size=self.args.chunk_size,
                        min_col_len=self.args.min_col_len,
                        col_len_multiplier=self.args.col_len_multiplier,
                    )

                    if self.args.after_insert:
                        for query in self.args.after_insert.split(self.args.sql_delimiter):
                            self.connection.exec_driver_sql(query)

                # Output SQL statements
                else:
                    statement = table.to_sql_create_statement(
                        table_name,
                        dialect=self.args.dialect,
                        db_schema=self.args.db_schema,
                        constraints=not self.args.no_constraints,
                        unique_constraint=self.unique_constraint,
                    )

                    self.output_file.write(f'{statement}\n')

        if self.connection:
            if self.args.queries:
                queries = []
                for query in self.args.queries:
                    if os.path.exists(query):
                        with open(query) as f:
                            query = f.read()
                    queries += query.split(self.args.sql_delimiter)

                # Execute the specified SQL queries.
                rows = None

                for query in queries:
                    if query.strip():
                        rows = self.connection.exec_driver_sql(query)

                # Output the result of the last query as CSV
                if rows.returns_rows:
                    output = agate.csv.writer(self.output_file, **self.writer_kwargs)
                    output.writerow(rows._metadata.keys)
                    for row in rows:
                        output.writerow(row)

            transaction.commit()


def launch_new_instance():
    utility = CSVSQL()
    utility.run()


if __name__ == '__main__':
    launch_new_instance()
